dhvani/man/dhvani.1
.\" '\" te
.TH "Dhvani" "1" "30 June 2008" "Santhosh Thottingal" ""
.SH "NAME"
dhvani \- Indian Language Text To Speech System. 
.SH "SYNOPSIS"
.PP 

.SH ""
.PP 
Dhvani is an Indian Language Text to Speech System. It can speak many Indian languages including Hindi, Panjabi,  Gujarati,  Bengali, Oriya, Telugu, Tamil, Kannada, Marathi and Malayalam. It takes text in unicode format and produces the speech. 
.PP 

\fBdhvani\fR [\fBoptions\fR] [\fBfile|text\fR]
.SH ""
.PP 
The following options are supported:
.sp
.ne 2
.mk
\fB\-\fBl\fR\fR
.in +9n
.rt
Set the language of the file. If this option is not set, dhvani will detect the language automatically. you need to give the language code with this option. You can also use \-\fB\-lang\fR to specify this option\&. One example is 

 dhvani \-l ml /path/to/malayalamfile.txt

language codes can be any of hi, ml, ka, ta, pa, gu, or, te, bn, mr. You can even use the standard format like hi_IN or ml_IN etc.

.sp
.sp 1
.in \-9n
.sp
.ne 2
.mk
\fB\-\fBs\fR\fR
.in +9n
.rt
Speed of the speech. You can also use \-\fB\-speed\fR to specify the option. It tell how much percentage the speed should be increased than the normal speed. \-\fB\-speed\fR option also does the same.

 dhvani \-s 10 /path/to/file.txt

.sp
.sp 1
.in \-9n
.sp
.ne 2
.mk
\fB\-\fBp\fR\fR
.in +9n
.rt
Pitch of the speech. You can also use \-\fB\-pitch\fR to specify the option. It tells by howmuch semitones  the pitch should increase or decrease. The value can be anything between \-60 to +60 semitones. You can also use \-\fB\-pitch\fR option.

 dhvani \-p 2 /path/to/file.txt


.sp
.sp 1
.in \-9n
.sp
.ne 2
.mk
\fB\-\fBo\fR\fR
.in +9n
.rt
Output file option. You can give one filename with this option to get the speech in an output file. You can also use \-\fB\-output\fR to specify this option\&.One example is 

 dhvani \-o /path/to/outputfile.ogg  /path/to/inputfile.txt 

The default output format is ogg format.

.sp
.sp 1
.in \-9n
.sp
.ne 2
.mk
\fB\-\fBi\fR\fR
.in +9n
.rt
Interactive mode. In this mode, user can input one sentence at prompt and hear the speech. Suitable for experimenting.  You can also use \-\fB\-format\fR option. Example:
 dhvani -i

You can even pipe the content of a text file to dhvani with this option. Example:
 dhvani -i < /path/of/the/file.txt

.sp
.sp 1
.in \-9n
.sp
.ne 2
.mk
\fB\-\fBt\fR\fR
.in +9n
.rt
Speak the given text given as the parameter.  You can also use \-\fB\-text\fR option. Example.
 dhvani -t "the text to speech"

.sp
.sp 1
.in \-9n
.sp
.ne 2
.mk
\fB\-\fBf\fR\fR
.in +9n
.rt
Output file format option. The default output format is ogg format. You can specify the format as wav or ogg using this switch You can also use \-\fB\-output\fR to specify this option\&.One example is 

 dhvani \-o /path/to/outputfile.wav \-f wav  /path/to/inputfile.txt 

or

 dhvani \-o /path/to/outputfile.ogg \-f ogg  /path/to/inputfile.txt 

The output file in wav fomrat will be a sound file in 16 bit Little Endian format with sampling rate of 16000 Hz To play this file use the following command.

 aplay \-f S16_LE \-r 16000 outputfile.wav


.sp
.sp 1
.in \-9n
.sp
.ne 2
.mk
\fB\-\fBd\fR\fR
.in +9n
.rt
This option will print all the supported languages.You can also use \-\fB\-list\fR to specify this option\&. Example usage:

 dhvani \-\-list
 
.sp
.sp 1
.in \-9n
.sp
.ne 2
.mk
\fB\-\fBh\fR\fR
.in +9n
.rt
To get the command help. You can also use \-\fB\-help\fR to specify this option\&.


.SH "Developers"
.PP 
 Dr. Ramesh Hariharan
.sp
 Santhosh Thottingal
.sp
 Rahul Bhalerao
.PP 

.SH "Bugs or Suggestions"
.PP 
 Visit http://dhvani.sourceforge.net and there is a mailing list for dhvani \- dhvani-devel@lists.sourceforge.net and there is an IRC channel also. 
 #dhvani @ irc.freenode.net
.PP 

.SH "License"
.PP 
Dhvani is licensed under GNU General Public License version 2 or any later version.
.PP 



dhvani/ChangeLog
version 0.94 :
	* Marathi support
	* Option to save the speech in ogg format
	* Option to modify the pitch and tempo of the speech
	* API fixes
	* Better debug support
	* Interactive mode option
	* Bug fixes and code restructuring

version 0.93 :
	* Tamil support Added

version 0.92 :
	* dhvani library is developed
	* Option to save the speech as a file
	* Option to pass the text from stdin
	*  many bugfixes

version 0.91 :
	* Added the following language support
		- Telugu
		- Gujarati
		- Bengali
		- Oriya
		- Panjabi
	* Bug fixes

version 0.81 :
	* Moved to single executable format from the client server method
	* Bug fixes

version 0.71 :
	* changed the sound system to ALSA from /dev/dsp direct access
	* Added Malayalam Support


dhvani/src/dhvani.c
/* dhvani.c
 *
 * Copyright (C) 2007-2008
 *  Santhosh Thottingal <santhosh.thottingal@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* This version of the command-line speak program uses the libdhvani shared  library */

#include<stdio.h>
#include<stdlib.h>
#include<unistd.h>
#include<string.h>
#include<sys/types.h>
#include <getopt.h>
#include "dhvani_lib.h"
#include "debug.h"
#include "languages.h"


static const char *help_text =
        "Dhvani Indian Language Text To Speech System\n"
        "Usage: dhvani [options] [file|text]\n"
        "[options] is any of the following \n"
        "  -l, --lang <languagecode>\t Set the language of the file. If this option is not set, dhvani will detect the language automatically.\n"
        /*"  -p, --phonetic\t Read Phonetic code.\n"*/
        "  -t, --text\t Say a text.\n"
        "  -d, --list\t Print the list of supported languages.\n"
        "  -s=n, --speed=n : Change sound tempo by n percents  (n=-95..+5000 %).\n"
        "  -p=n, --pitch=n : Change sound pitch by n semitones (n=-60..+60 semitones).\n"
        "  -o, --output <file>\t Save the speech to a file.\n"
        "  -f, --format <ogg|wav>\t Format of the generated speech file while writing to a file. Default is ogg.\n"
        "  -v, --version\t Print the version.\n"
        "  -i, --stdin\t Read the content from stdin. Interactive mode.\n"
        "  -h, --help\t Print this help.\n"
        "File should be in UTF-8 format.\n"
        "Report bugs/suggestions to Santhosh Thottingal <santhosh.thottingal@gmail.com>.\n";

/*
Print the Dhvani command line help message.
 */
void
print_usage()
{
        dhvani_info("Dhvani Indian Language Text To Speech System: %s\n%s",
                dhvani_Info(), help_text);
}

int
main(int argc, char *argv[])
{
        static struct option long_options[] = {
                /* These options set a flag. */
                /* These options don't set a flag.
                   We distinguish them by their indices. */
                {"help", no_argument, 0, 'h'},
                {"version", no_argument, 0, 'v'},
                {"lang", optional_argument, 0, 'l'},
                {"speed", optional_argument, 0, 's'},
                {"output", optional_argument, 0, 'o'},
                {"format", optional_argument, 0, 'f'},
                {"phonetic", optional_argument, 0, 'k'},
                {"pitch", optional_argument, 0, 'p'},
                {"text", optional_argument, 0, 't'},
                {"list", no_argument, 0, 'd'},
                {"debug", no_argument, 0, 'X'},
                {"stdin", no_argument, 0, 'i'},
                {0, 0, 0, 0}
        };
        FILE *fd = 0;
        dhvani_options *options;
        int text_option = 0;
        char *text = NULL;
        int option_index = 0;
        int flag_stdin = 0;
        int argument_character;
        int max = 1000;
        char *stdin_text = (char *) malloc(max);

	options = dhvani_init();
        options->isPhonetic = 0;
        options->speech_to_file = 0;
        options->pitch = 0.0;
        options->tempo = 10;
        options->rate = 16000;
        options->output_file_format = DHVANI_OGG_FORMAT;
        if (argc == 1) {
                print_usage();
                exit(0);
        }

	
        while (1) {
                argument_character =
                        getopt_long(argc, argv, "idhvp:l:o:s:f:t:X", long_options, &option_index);
                /* Detect the end of the options. */
                if (argument_character == -1)
                        break;
                switch (argument_character) {
                case 'd':
                        printf("%s\n", dhvani_ListLanguage());
                        exit(0);
                        break;

                case 'h':
                        print_usage();
                        exit(0);
                        break;

                case 'v':
                        printf("%s\n", dhvani_Info());
                        break;

                case 'k':
                        fd = fopen(optarg, "r");
                        options->isPhonetic = 1;
                        break;

                case 'p':
                        options->pitch = atoi(optarg);
                        break;

                case 't':
                        text = optarg;
                        options->isPhonetic = 0;
                        text_option = 1;
                        break;

                case 'o':
                        options->speech_to_file = 1;
                        /*Default output format is ogg*/
                        options->output_file_name = optarg;
                        break;

                case 'l':
                        options->language = get_language_code(optarg);
                        break;

                case 's':
                        options->tempo = atoi(optarg);
                        break;
                case 'f':
                        if (strcmp(optarg, "WAV") == 0 || strcmp(optarg, "wav") == 0 || strcmp(optarg, "WAVE") == 0 || strcmp(optarg, "wave") == 0) {
                                options->output_file_format = DHVANI_WAV_FORMAT;
                        }
                        if (strcmp(optarg, "OGG") == 0 || strcmp(optarg, "ogg") == 0) {
                                options->output_file_format = DHVANI_OGG_FORMAT;
                        }
                        break;

                case 'i':
                        flag_stdin = 1;
                        break;

                default:
                        exit(0);
                }
        }
        if ((!flag_stdin ) && (options->isPhonetic == 0) && (!text_option)) {
                fd = fopen(argv[argc - 1], "r");
                if (fd == NULL) {
                        dhvani_error("Could not open file %s\n", argv[argc - 1]);
                        exit(0);
                }
        }
        if (options->isPhonetic == 1) {
                /* Phonetic file as input*/
                dhvani_debug("Phonetic mode \n");
               /* dhvani_init(&options); */
                dhvani_speak_phonetic_file(fd);
        } else if (text_option == 0 && flag_stdin == 0) {
                dhvani_debug("Text file mode \n");
               /* dhvani_init(&options); */
                /* Speak the given file*/
                dhvani_speak_file(fd, options);
        } else if (!options->isPhonetic && text_option == 1) {
                /* Now option is o speak the given text*/
                dhvani_debug("Text mode \n");
                /* dhvani_init(&options); */
                dhvani_say(text, options);
        } else if (flag_stdin == 1) {
                /* line by line input on stdin. First initialize */
                dhvani_debug("Interactive mode \n");
                /* dhvani_init(&options); */
                printf("dhvani interactive mode.Give the string to speak at the prompt");
                printf("\ndhvani>");
		fflush(stdin);
		fflush(stdout);
                while (fgets(stdin_text, max, stdin) != NULL) {
                        /*say it */
                        dhvani_say(stdin_text, options);
                        /*display the prompt */
                        printf("\ndhvani>");
                }
        }
        if (fd != NULL) {
                fclose(fd);
        }
        return 0;
}
